elephant/gpfa_src/gpfa_util.py

```diff
@@ -0,0 +1,579 @@
+
+    return seq
+
+
+def cut_trials(seq_in, seg_length=20):
+    """
+    Extracts trial segments that are all of the same length.  Uses
+    overlapping segments if trial length is not integer multiple
+    of segment length.  Ignores trials with length shorter than
+    one segment length.
+
+    Parameters
+    ----------
+
+    seq_in
+        data structure, whose nth entry (corresponding to
+        the nth experimental trial) has fields
+            * trialId: unique trial identifier
+            * T: (1 x 1) number of timesteps in trial
+            * y: (yDim x T) neural data
+
+    seg_length : int
+        length of segments to extract, in number of timesteps. If infinite,
+        entire trials are extracted, i.e., no segmenting.
+        Default is 20
+
+
+    Returns
+    -------
+
+    seqOut
+        data structure, whose nth entry (corresponding to
+        the nth experimental trial) has fields
+            * trialId: identifier of trial from which segment was taken
+            * segId: segment identifier within trial
+            * T: (1 x 1) number of timesteps in segment
+            * y: (yDim x T) neural data
+    """
+    assert seg_length > 0, "At least 1 extracted trial must be returned"
+    if np.isinf(seg_length):
+        seqOut = seq_in
+        return seqOut
+
+    dtype_seqOut = [('trialId', np.int), ('segId', np.int), ('T', np.int),
+                    ('y', np.object)]
+    seqOut_buff = []
+    for n, seqIn_n in enumerate(seq_in):
+        T = seqIn_n['T']
+
+        # Skip trials that are shorter than segLength
+        if T < seg_length:
+            warnings.warn('trialId {0:4d} shorter than one segLength...'
+                          'skipping'.format(seqIn_n['trialId']))
+            continue
+
+        numSeg = np.int(np.ceil(float(T) / seg_length))
+
+        # Randomize the sizes of overlaps
+        if numSeg == 1:
+            cumOL = np.array([0, ])
+        else:
+            totalOL = (seg_length * numSeg) - T
+            probs = np.ones(numSeg - 1, np.float) / (numSeg - 1)
+            randOL = np.random.multinomial(totalOL, probs)
+            cumOL = np.hstack([0, np.cumsum(randOL)])
+
+        seg = np.empty(numSeg, dtype_seqOut)
+        seg['trialId'] = seqIn_n['trialId']
+        seg['T'] = seg_length
+
+        for s, seg_s in enumerate(seg):
+            tStart = seg_length * s - cumOL[s]
+
+            seg_s['segId'] = s
+            seg_s['y'] = seqIn_n['y'][:, tStart:tStart + seg_length]
+
+        seqOut_buff.append(seg)
+
+    if len(seqOut_buff) > 0:
+        seqOut = np.hstack(seqOut_buff)
+    else:
+        seqOut = np.empty(0, dtype_seqOut)
+
+    return seqOut
+
+
+def rdiv(a, b):
+    """
+    Returns the solution to x b = a. Equivalent to MATLAB right matrix
+    division: a / b
+    """
+    return np.linalg.solve(b.T, a.T).T
+
+
+def logdet(A):
+    """
+    log(det(A)) where A is positive-definite.
+    This is faster and more stable than using log(det(A)).
```

Comment:
Is this equivalent in python/numpy?

Reply:
This is equivalent to `np.log(np.linalg.det(A))` and, what is more important, numerically stable. However, this function is actually _slower_ than numpy's version. Maybe because I've installed MKL.
A good catch. I'll add a special test.
![Figure_1](https://user-images.githubusercontent.com/7688337/60604234-80621b80-9db7-11e9-8940-dadb2b5a27de.png)